src/shape_transform_descriptor.cpp

```diff
@@ -0,0 +1,836 @@
+}
+
+// Search all subdimensions and return the subdimensions vector, an iterator
+// to the subdimension found and an optional iterator to the previous
+// subdimension if available.
+template <class Predicate>
+static auto find_subdimension(shape_transform_descriptor& td, Predicate p)
+{
+    dimension* prev_dim = nullptr;
+    for(auto& d : td.dimensions)
+    {
+        auto it = std::find_if(d.subdimensions.begin(), d.subdimensions.end(), p);
+        if(it != d.subdimensions.end())
+        {
+            decltype(std::make_optional(it)) prev = nullopt;
+            if(it == d.subdimensions.begin())
+            {
+                if(prev_dim != nullptr and not prev_dim->subdimensions.empty())
+                {
+                    prev = std::prev(prev_dim->subdimensions.end());
+                }
+            }
+            else
+            {
+                prev = std::prev(it);
+            }
+            return std::make_tuple(&d.subdimensions, it, prev);
+        }
+        prev_dim = &d;
+    }
+    MIGRAPHX_THROW("Searching for non-existent subdimension");
+}
+
+static bool is_broadcast_dim(const dimension& d)
+{
+    if(d.len() == 1)
+        return false;
+    assert(not d.subdimensions.empty());
+    if(d.subdimensions.size() != 1)
+        return false;
+    const auto& sub = d.subdimensions.front();
+    return sub.axis.empty();
+}
+static bool missing_leading_axis(const dimension& d)
+{
+    if(d.subdimensions.empty())
+        return true;
+    const auto& sub = d.subdimensions.front();
+    return sub.axis.empty();
+}
+
+static void set_broadcast_dim(dimension& d, std::size_t axis)
+{
+    if(d.subdimensions.empty())
+        d.subdimensions.push_back({1, {axis}});
+    else
+        d.subdimensions.front().hidden_axis = axis;
+}
+
+// Group all axes into a map with a key of the axis and the value is vector of
+// all subdimensions that have that axis.
+static std::map<std::size_t, std::vector<dimension::sub*>>
+group_axes(std::vector<dimension>& dimensions)
+{
+    std::map<std::size_t, std::vector<dimension::sub*>> axes_map;
+    for(auto& d : dimensions)
+    {
+        for(auto& s : d.subdimensions)
+        {
+            if(s.axis.empty())
+                continue;
+            axes_map[s.axis.front()].push_back(&s);
+        }
+    }
+    return axes_map;
+}
+
+// Renumber all axes while preserving the order of the axes
+static void renumber_axes(std::map<std::size_t, std::vector<dimension::sub*>>& axes_map)
+{
+    for(auto&& p : axes_map)
+    {
+        const auto& axis = p.first;
+        auto& subs       = p.second;
+        if(subs.size() == 1)
+        {
+            subs[0]->axis = {axis};
+        }
+        else
+        {
+            std::sort(subs.begin(), subs.end(), by(std::less<>{}, [](const dimension::sub* s) {
+                          return s->axis;
+                      }));
+            for(std::size_t i : range(subs.size()))
+                subs[i]->axis = {axis, i};
```

Comment:
how does this work for the multiple split case (ie. same dim got split multiple times)? Eg. you have 3 subdims like 
`[[len=x, axis={0,2,0}]], [[len=y, axis={0,2,1}]], [[len=z, axis={0,2,2}]]`
and then simplify is able to simplify 2 of them st. you have remaining:
`[[len=x*y, axis={0,2,1}]], [[len=z, axis={0,2,2}]]`
If I understand correctly, renumbering should make those axis {0,2,0} and {0,2,1} right?

Reply:
Further along the same train of thought, if simplify finds that all of them are adjacent and gives
`[[len=x*y*z, axis={0,2,2}]]`
Then is there a way somewhere to remove that last value from the axis vector entirely?

Reply:
>  and then simplify is able to simplify 2 of them st. you have remaining:

Simplify is able to simplify all three of them into the same dim, which will be `{0, 2, 2}`. 

>  If I understand correctly, renumbering should make those axis {0,2,0} and {0,2,1} right?

Its kind of missing what the other axes are, but lets assume the other axes couldn't be simplified and are: `{0, 0, 0}`, `{0, 0, 1}`, `{0, 1, 0}`, `{0, 1, 1}`. So renumbering will change the all axes to:

* `{0, 0, 0}` -> `{0, 0}`
* `{0, 0, 1}` -> `{0, 1}`
* `{0, 1, 0}` -> `{0, 2}`
* `{0, 1, 1}` -> `{0, 3}`
* `{0, 2, 2}` -> `{0, 4}`

Reply:
>  Then is there a way somewhere to remove that last value from the axis vector entirely?

If all them got simplified then it will just set the axis to `{0}`. Thats what the condition above does(ie under `if(subs.size() == 1)`).